File: 3653-maximum-subarray-sum-with-length-divisible-by-k/maximum-subarray-sum-with-length-divisible-by-k.py
class Solution:
    def maxSubarraySum(self, nums: List[int], k: int) -> int:
        n = len(nums)
        dp = [inf] * (k - 1) + [0]
        pre, ans = 0, -inf

        for key, val in enumerate(nums):
            pre += val
            nk = key % k
            ans = max(ans, pre - dp[nk])
            dp[nk] = min(pre, dp[nk])

        return ans

File: 3653-maximum-subarray-sum-with-length-divisible-by-k/maximum-subarray-sum-with-length-divisible-by-k.cs
public class Solution {
    public long MaxSubarraySum(int[] nums, int k) {
        int n = nums.Length;
        long[] dp = new long[k];
        long sum = 0, ans = long.MinValue;

        for(int i = 0; i < n; i++) {
            sum += nums[i];
            
            if (i + 1 >= k) {
                ans = Math.Max(ans, sum - dp[(i + 1) % k]);
                dp[(i + 1) % k] = Math.Min(dp[(i + 1) % k], sum);
            } else {
                dp[(i + 1) % k] = sum; 
            }
        }

        return ans;
    }
}